### ISSUE TITLE
Tutorial on environment setup and question about Cora node classification accuracy (~70%)

### ISSUE THREAD

**Issue by shwangtangjun** [zhouc20/LatentGraphDiffusion#18]
Since there are already many issues related to environment configuration, e.g. [2](https://github.com/zhouc20/LatentGraphDiffusion/issues/2), [3](https://github.com/zhouc20/LatentGraphDiffusion/issues/3), [4](https://github.com/zhouc20/LatentGraphDiffusion/issues/4), [7](https://github.com/zhouc20/LatentGraphDiffusion/issues/7), [9](https://github.com/zhouc20/LatentGraphDiffusion/issues/9), [10](https://github.com/zhouc20/LatentGraphDiffusion/issues/10), [12](https://github.com/zhouc20/LatentGraphDiffusion/issues/12), [13](https://github.com/zhouc20/LatentGraphDiffusion/issues/13), [14](https://github.com/zhouc20/LatentGraphDiffusion/issues/14), and I personally spent almost two days resolving similar problems, I would like to share a complete environment setup that finally allows me to run the experiments from scratch. So far, I have only tested Cora node classification.

### 1. Bug in `pyg==2.0.4` (the version suggested in README)
The line
~~~python
from torch_geometric.graphgym.imports import LightningModule
~~~
throws an error. This seems related to:  https://github.com/pyg-team/pytorch_geometric/issues/4829. I did not find a direct fix.

### 2. Massive missing attribute errors

For example: `name_tag`, `run_multiple_splits`, etc.  This is mentioned repeatedly in many earlier issues.

## How I resolved them

- **Problem 1**: install the latest torch-geometric (as of Feb 4, 2026).
- **Problem 2**: add an `__init__.py` file under the `/lgd` directory.

### 0. Create environment

~~~bash
conda create -n lgd python=3.12
conda activate lgd
~~~

### 1. Install PyTorch

~~~bash
pip install torch==2.8.0 torchvision==0.23.0 --index-url https://download.pytorch.org/whl/cu129
~~~

(I choose torch 2.8 because it is currently the newest version supported by torch-geometric. Choose the wheel URL according to your CUDA version.)

### 2. Install PyG

~~~bash
pip install torch_geometric
pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.8.0+cu129.html
~~~

### 3. Install other dependencies

~~~bash
pip install yacs lightning einops performer_pytorch ogb opt_einsum seaborn fcd_torch rdkit EMD-signal tensorboardX
~~~

### 4. **Add `/lgd/__init__.py`** (crucial)

Create a file `lgd/__init__.py` with:

~~~python
from .asset import *
from .config import *
from .ddpm import *
from .encoder import *
from .head import *
from .layer import *
from .loader import *
from .loss import *
from .model import *
from .optimizer import *
from .train import *
from .transform import *
~~~

After this, the missing-attribute errors should disappear.

I honestly do not understand why this critical step has been repeatedly overlooked when answering related issues. Continuously telling users to “go to the [GraphGPS](https://github.com/rampasek/GraphGPS) repository and find the answer by themselves”, without explicitly stating this root cause, is **irresponsible** and wastes a huge amount of others' time.

### 5. Remove deprecated import

In `utils.py`, delete the following line (not supported by Python 3):

~~~python
import imp
~~~

### 6. (Optional) Checkpoint path

In `cora-node-diffusion.yaml`, around line 169, update the checkpoint path (change 2->0) as stated in the README:

> Remember to change the file path of the checkpoint of the autoencoder in `diffusion.first_stage_config`.

## Run commands

~~~bash
python pretrain.py --cfg cfg/cora-node-encoder.yaml --repeat 1 wandb.use False
python train_diffusion.py --cfg cfg/cora-node-diffusion.yaml --repeat 1 wandb.use False
~~~

At this point, everything runs successfully.

## Question: Why is accuracy only ~70%?

My results are very similar to issue [6](https://github.com/zhouc20/LatentGraphDiffusion/issues/6).

- Encoder pretraining: `accuracy ≈ 0.72`
- Diffusion training: `accuracy ≈ 0.71 ~ 0.73`

<details>
<summary>Logs: Encoder pretraining (pretrain.py)</summary>

~~~text
Avg time per epoch: 0.35s
Total train loop time: 0.05h
Task done, results saved in results\cora-node-encoder\0
369
{'epoch': 369, 'time_epoch': 0.10909, 'loss': 1.19012773, 'lr': 0, 'params': 779731, 'time_iter': 0.10909, 'accuracy': 0.718, 'f1': 0.71034, 'auc': 0.89265, 'loss_labeled': 0.7569039463996887, 'metric_labeled': 0.969}
{'epoch': 369, 'time_epoch': 0.103, 'eta': 15.50112, 'eta_hours': 0.00431, 'loss': 0.7154724, 'lr': 0.00019493, 'params': 779731, 'time_iter': 0.103, 'accuracy': 1.0, 'f1': 1.0, 'auc': 1.0}
{'epoch': 369, 'time_epoch': 0.0967, 'loss': 1.21461749, 'lr': 0, 'params': 779731, 'time_iter': 0.0967, 'accuracy': 0.726, 'f1': 0.71478, 'auc': 0.87976, 'loss_labeled': 0.7245136499404907, 'metric_labeled': 0.992}
~~~
</details>

<details>
<summary>Logs: Diffusion training (train_diffusion.py)</summary>

~~~text
Avg time per epoch: 2.97s
Total train loop time: 0.17h
Task done, results saved in results\cora-node-diffusion\0
114
{'epoch': 114, 'time_epoch': 6.66575, 'loss': 1.24257016, 'lr': 0, 'params': 2418791, 'time_iter': 6.66575, 'accuracy': 0.725, 'f1': 0.71789, 'auc': 0.89461}
{'epoch': 114, 'time_epoch': 0.10843, 'eta': 9.12299, 'eta_hours': 0.00253, 'loss': 1.18687534, 'lr': 4.651e-05, 'params': 2418791, 'time_iter': 0.10843, 'accuracy': 0.97857, 'f1': 0.97853, 'auc': 0.99601, 'loss_node': 0.31367790699005127, 'loss_task': 0.8731974363327026}
{'epoch': 114, 'time_epoch': 6.7816, 'loss': 1.26147759, 'lr': 0, 'params': 2418791, 'time_iter': 6.7816, 'accuracy': 0.714, 'f1': 0.69104, 'auc': 0.87875}
~~~
</details>

**Important note:** I have **not changed any code from the latest commit**, except adding `lgd/__init__.py`.

It is hard to believe that differences in torch / torch-geometric versions alone would cause a drop from **93.91% (paper)** to **~70%**.

Could the authors clarify:

- Is there any missing step not documented?
- Are the released configs guaranteed to reproduce the reported numbers?

## A suggestion

Last but not least, I understand that this repository is inherited from [GraphGPS](https://github.com/rampasek/GraphGPS), which itself is built on top of GraphGym. However, building the entire codebase on top of GraphGym still makes it extremely difficult for others to dive into the code and identify what truly matters.

I hereby propose that future researchers should at least provide a **standalone minimal working example** to demonstrate their core contribution, rather than hiding all components inside a huge and highly abstracted repository.

**Comment by zhouc20** [comment]
Hi,

Thank you for the complete tutorial. 

1. Regarding environment: Indeed I forgot to upload the __init__.py which induces the import errors. The reason I refer to previous GraphGPS repo is because I can no longer access my original environment, and I created my own environment following their setup. I believe referring to previous repo is extremely common and could address the issues. However, I sort of disagree with your opinion that one should always provide a standalone repo -- building upon existing high-quality codebase is not only more efficient and may lead to better performance, but also clearly compare your own version with the baselines, thus clearly differentiating your new contribution; otherwise, the comparison with baselines may not be fair due to different experimental setups and implementation details.

2. Regarding results on Cora: I used to replicate the results using new environments, and it indeed leads to significantly worse results. Here I attached the experiment logs, checkpoints and configs of my original experiments on Cora, which is consistent with the numbers reported in the paper (https://drive.google.com/file/d/14evkHOMKsq4rU1iYbCsER6vr9coeE4ij/view?usp=sharing). Note that these results are produced using the old pyg versions, and you may encounter errors while loading the checkpoints due to implementation difference of pyg. Regarding your case, it is not the fault of the diffusion stage: you already get a very low score in your encoder, which accounts for the low accuracy of diffusion. You can also try other configurations or architectures to get a better encoder (as reported in many other papers), and make sure to use a good encoder before proceeding to the diffusion stage. After all, learning a good encoder is not the focus of this paper, and you can use any existing discriminative model as the encoder.

(P.S. throughout my career while the graph learning benchmarks are universally with low quality (https://arxiv.org/abs/2502.14546), I believe node classification datasets like Cora are especially noisy and lack practical meanings. I would strongly suggest switching to other datasets for more meaningful research. Me and my research group do not care about node classification at all - the only reason the paper includes this task comes from one of the reviewers' suggestion. Nevertheless, all the reported numbers are reproducible anyways - see above.)
